smartdevicelink/sdl_java_suite
# Feature/issue 811 broaden choiceset uniqueness

Fixes #811 

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
- [x] I have tested Android, Java SE, and Java EE

#### Unit Tests
Unit tests have been updated for changes

#### Core Tests
To Test Choice Sets
in SdlService of HelloSdl preload a choiceSet that contains choiceCells that contains choiceCells where some choiceCells have the same primaryText and all choiceCells have different secondaryText, example below.
```
ChoiceCell cell1 = new ChoiceCell("McDonalds", "1 mile away", null, null, null, null);
ChoiceCell cell2 = new ChoiceCell("McDonalds", "2 mile away", null, null, null, null);
ChoiceCell cell3 = new ChoiceCell("Starbucks", "3 mile away", null, null, null, null);
ChoiceCell cell4 = new ChoiceCell("McDonalds", "4 mile away", null, null, null, null);
ChoiceCell cell5 = new ChoiceCell("Starbucks", "5 mile away", null, null, null, null);
ChoiceCell cell6 = new ChoiceCell("Meijer", "6 mile away", null, null, null, null);
ChoiceCell cell7 = new ChoiceCell("McDonalds", "7 mile away", null, null, null, null);
ChoiceCell cell8 = new ChoiceCell("Speedway", "8 mile away", null, null, null, null);
ChoiceCell cell9 = new ChoiceCell("McDonalds", "9 mile away", null, null, null, null);
ChoiceCell cell10 = new ChoiceCell("McDonalds", "10 mile away", null, null, null, null);
ChoiceCell cell11 = new ChoiceCell("McDonalds", "11 mile away", null, null, null, null);
ChoiceCell cell12 = new ChoiceCell("Starbucks", "12 mile away", null, null, null, null);
ChoiceCell cell13 = new ChoiceCell("Starbucks", "13 mile away", null, null, null, null);
ChoiceCell cell14 = new ChoiceCell("Speedway", "14 mile away", null, null, null, null);
ChoiceCell cell15 = new ChoiceCell("McDonalds", "15 mile away", null, null, null, null);

choiceCellList = new ArrayList<>(Arrays.asList(cell1, cell2, cell3, cell4, cell5, cell6, cell7, cell8, cell9, cell10, cell11, cell12, cell13, cell14, cell15));
sdlManager.getScreenManager().preloadChoices(choiceCellList, null);
```

Connect the HelloSdl app to an instance of core where SdlMsgVersion is 7.0 and verify that all choiceCells appear with a name that has "(X)" appended at the end where "X" is the number of times this text havs been used (The first instance of a name will not have (1) appended after), example below.
```
McDonalds
McDonalds (2)
Starbucks
McDonalds (3)
Starbucks (2)
Meijer
McDonalds (4)
Speedway
McDonalds (5)
McDonalds (6)
McDonalds (7)
Starbucks (3)
Starbucks (4)
Speedway (2)
McDonalds (8)
```

Connect the HelloSdl app to an instance of core where SdlMsgVersion is 7.1 and verify that all choiceCells appear with the name that was given in SdlService, example below.
```
McDonalds
McDonalds
Starbucks
McDonalds
Starbucks
Meijer
McDonalds
Speedway
McDonalds
McDonalds
McDonalds
Starbucks
Starbucks
Speedway
McDonalds
```

To Test Menu Manager
in SdlService of HelloSdl set the menu that contains MenuCells where some MenuCells have the same primaryTitlebut different icons, example below.
```
MenuCell subCell1 = new MenuCell("Name 2", image1, null, new MenuSelectionListener() {});
MenuCell subCell2 = new MenuCell("Name 2", image2, null, new MenuSelectionListener() {});
MenuCell mainCell1 = new MenuCell("Name 1", MenuLayout.LIST, image1, Arrays.asList(subCell1, subCell2));
MenuCell subCell3 = new MenuCell("Name 3", image1, null, new MenuSelectionListener() {});
MenuCell subCell4 = new MenuCell("Name 3", image2, null, new MenuSelectionListener() {});
MenuCell mainCell2 = new MenuCell("Name 1", MenuLayout.LIST, image2, Arrays.asList(subCell3, subCell4));
MenuCell mainCell3 = new MenuCell("Test 1", image1, null, new MenuSelectionListener() {});
MenuCell mainCell4 = new MenuCell("Test 1", image2, null, new MenuSelectionListener() {});

sdlManager.getScreenManager().setMenu(Arrays.asList(mainCell1, mainCell2, mainCell3, mainCell4));
```

Connect the HelloSdl app to an instance of core where SdlMsgVersion is 7.0 and verify that all MenuCells appear with a name that has "(X)" appended at the end where "X" is the number of times this text havs been used, example below (Top row is top level menu, following rows are sub menus).
| Name 1 | Name 1 (2) | Test 1 | Test 1 (2) | 
|---------|------------|---|---|
| Name 2 | Name 3 | | |
| Name 2 (2) | Name 3 (3) | | |

Connect the HelloSdl app to an instance of core where SdlMsgVersion is 7.1 and verify that all MenuCells appear with the name that was given in SdlService, example below (Top row is top level menu, following rows are sub menus).
| Name 1 | Name 1 | Test 1 | Test 1 |
|---------|--------|---|---|
| Name 2 | Name 3 | | |
| Name 2 | Name 3 | | |

Core version / branch / commit hash / module tested against: [Core PR](https://github.com/smartdevicelink/sdl_core/pull/3607)

### Summary
This PR will add a `uniqueText` property to the `ChoiceCell` class that will be used when `SdlMsgVersion` is older than 7.1, in this case core does not allow for ChoiceCells to have the same primary text. `addUniqueNamesToCells` has been added to `BaseChoiceSetManager` to go through the list of choices to be preloaded and update the uniqueText property to ensure all ChoiceCells have a different menu name to be displayed. If the SdlMsgVersion is 7.1 or newer core will support identical menu names so the ChoiceCell primary text will be used and the uniqieText will not be set. in the `PreloadChoicesOperation` we also check the `SdlMsgVersion` to decide if the menuName of the `CreateInteractionChoiceSet` will use the `uniqueText` or the primary text.

Currently on Develop when a duplicate voice command is set on a menuCell and a subCell then the menuCell is uploaded with the voiceCommand and the subCell is not uploaded

With these changes when a duplicate voice command is set on a menuCell and a subCell then the menu will not be uploaded

### CLA
